Separate move path tracking between borrowck and drop elaboration.

cjgillot opened the pull request:
The primary goal of this PR is to skip creating a `MovePathIndex` for path that do not need dropping in drop elaboration.

The 2 first commits are cleanups.

The next 2 commits displace `move` errors from move-path builder to borrowck. Move-path builder keeps the same logic, but does not carry error information any more.

The remaining commits allow to filter `MovePathIndex` creation according to types. This is used in drop elaboration, to avoid computing dataflow for paths that do not need dropping.

cjgillot committed:
Do not report errors from move path builder.

cjgillot committed:
Remove is_terminal_path.

That information is redundant. Is the path is terminal, `first_child` will
already be `None`.

cjgillot committed:
Stop computing error info in move path builder.

cjgillot committed:
Parametrize gather_moves by filter.

cjgillot committed:
Do not create move paths that do not need dropping.

cjgillot commented:
@bors try @rust-timer queue

rust-timer commented:
Awaiting bors try build completion.

@rustbot label: +S-waiting-on-perf

<!-- rust-timer: temporary -->

bors commented:
:hourglass: Trying commit 9e7067bdbe171d06675752603d0bc9d21430d977 with merge 6e61e35c7511f1da2c2379bae4b5d2c214093e77...
<!-- homu: {"type":"TryBuildStarted","head_sha":"9e7067bdbe171d06675752603d0bc9d21430d977","merge_sha":"6e61e35c7511f1da2c2379bae4b5d2c214093e77"} -->

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/6366007864/job/17283962234)
Build commit: 6e61e35c7511f1da2c2379bae4b5d2c214093e77 (`6e61e35c7511f1da2c2379bae4b5d2c214093e77`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/6366007864/job/17283962234"},"merge_sha":"6e61e35c7511f1da2c2379bae4b5d2c214093e77"} -->

rust-timer commented:
Queued 6e61e35c7511f1da2c2379bae4b5d2c214093e77 with parent 2f89c414f2e3a05f104012b860c27338258f5382, future [comparison URL](https://perf.rust-lang.org/compare.html?start=2f89c414f2e3a05f104012b860c27338258f5382&end=6e61e35c7511f1da2c2379bae4b5d2c214093e77).
There are currently 2 preceding artifacts in the [queue](https://perf.rust-lang.org/status.html).
It will probably take at least ~3.6 hours until the benchmark run finishes.
<!-- rust-timer: temporary -->

rust-timer commented:
Finished benchmarking commit (6e61e35c7511f1da2c2379bae4b5d2c214093e77): [comparison URL](https://perf.rust-lang.org/compare.html?start=2f89c414f2e3a05f104012b860c27338258f5382&end=6e61e35c7511f1da2c2379bae4b5d2c214093e77&stat=instructions:u).

### Overall result: ✅ improvements - no action needed


Benchmarking this pull request likely means that it is perf-sensitive, so we're automatically marking it as not fit for rolling up. While you can manually mark this PR as fit for rollup, we strongly recommend not doing so since this PR may lead to changes in compiler perf.

@bors rollup=never
@rustbot label: -S-waiting-on-perf -perf-regression


### [Instruction count](https://perf.rust-lang.org/compare.html?start=2f89c414f2e3a05f104012b860c27338258f5382&end=6e61e35c7511f1da2c2379bae4b5d2c214093e77&stat=instructions:u)
This is a highly reliable metric that was used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -0.6% | [-1.4%, -0.3%] | 61    |
| Improvements ✅ <br /> (secondary) | -0.9% | [-1.6%, -0.2%] | 32    |
| All ❌✅ (primary)                 | -0.6% | [-1.4%, -0.3%] | 61    |

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=2f89c414f2e3a05f104012b860c27338258f5382&end=6e61e35c7511f1da2c2379bae4b5d2c214093e77&stat=max-rss)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 5.1%  | [5.1%, 5.1%]   | 1     |
| Regressions ❌ <br /> (secondary)  | 0.7%  | [0.7%, 0.8%]   | 4     |
| Improvements ✅ <br /> (primary)   | -3.6% | [-5.4%, -2.5%] | 3     |
| Improvements ✅ <br /> (secondary) | -1.4% | [-2.3%, -0.7%] | 4     |
| All ❌✅ (primary)                 | -1.4% | [-5.4%, 5.1%]  | 4     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=2f89c414f2e3a05f104012b860c27338258f5382&end=6e61e35c7511f1da2c2379bae4b5d2c214093e77&stat=cycles:u)
This benchmark run did not return any relevant results for this metric.

### [Binary size](https://perf.rust-lang.org/compare.html?start=2f89c414f2e3a05f104012b860c27338258f5382&end=6e61e35c7511f1da2c2379bae4b5d2c214093e77&stat=size:linked_artifact)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 0.0%  | [0.0%, 0.1%]   | 3     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -0.0% | [-0.0%, -0.0%] | 1     |
| Improvements ✅ <br /> (secondary) | -     | -              | 0     |
| All ❌✅ (primary)                 | 0.0%  | [-0.0%, 0.1%]  | 4     |
</details>

**Bootstrap**: 630.4s -> 631.994s (0.25%)
**Artifact size**: 317.38 MiB -> 317.23 MiB (-0.05%)

cjgillot committed:
Remove on_all_drop_children_bits.

As drop elaboration only tracks places that need dropping, is has become
equivalent to `on_all_children_bits`.

cjgillot commented:
r? compiler

petrochenkov commented:
I'm not very familiar with this code, but I worked with move paths several years ago when implementing unions, and I need to get familiar with it again for my work, so I'll review anyway.

petrochenkov commented on compiler/rustc_borrowck/src/diagnostics/move_errors.rs:
```diff
@@ -1,16 +1,50 @@
 use crate::MirBorrowckCtxt;
 
+#[derive(Debug)]
+pub enum IllegalMoveOriginKind<'tcx> {
```
In this commit it's hard to see that `IllegalMoveOriginKind` is not a new structure but is actually copypasted from elsewhere, and that the copy is identical.
Maybe it should initially be imported instead.

petrochenkov commented on compiler/rustc_borrowck/src/lib.rs:
```diff
@@ -1465,6 +1470,9 @@ impl<'cx, 'tcx> MirBorrowckCtxt<'cx, 'tcx> {
             }
             Operand::Move(place) => {
+                // Check if moving from this place makes sense.
+                self.check_movable_place(location, place);
```
Why is `check_movable_place` called from two places?
Can it be centralized to one place?

cjgillot replied:
We want to report move errors from promoted, but the different borrowck analyses are still run manually for them. This is the least-invasive way to keep the behaviour.
The "proper" way to unify both cases would be https://github.com/rust-lang/rust/pull/111693 or something like it.

petrochenkov commented on compiler/rustc_borrowck/src/lib.rs:
```diff
@@ -1586,6 +1594,142 @@ impl<'cx, 'tcx> MirBorrowckCtxt<'cx, 'tcx> {
+                // `OpaqueCast` only transmutes the type, so no moves there and
+                // `Downcast` only changes information about a `Place` without moving
+                // So it's safe to skip these.
+                ProjectionElem::OpaqueCast(_) | ProjectionElem::Downcast(_, _) => (),
```
Is it possible to share code between this function and `fn move_path_for` somehow?

In this commit it's hard to see that this logic is not new but is actually copied from elsewhere, and that the logic is identical in those two places.

petrochenkov replied:
Generic version, if that's not too ugly.

cjgillot replied:
I don't think keeping a common version is an interesting path forward. The borrowck version is tailored to the language behaviour in borrowck, ie user-facing. The mir_dataflow version could evolve into being more syntactic, and check less types.

cjgillot committed:
Avoid using a magic value for untracked locals.

cjgillot committed:
Give collect_drop_flags and elaborate_drops closer structure.

petrochenkov commented on compiler/rustc_mir_transform/src/elaborate_drops.rs:
```diff
@@ -324,80 +334,74 @@ impl<'b, 'tcx> ElaborateDropsCtxt<'b, 'tcx> {
-                                unwind,
-                                bb,
-                            )
+            // This place does not need dropping. It is not have an associated move-path, so the
```
```suggestion
            // This place does not need dropping. It does not have an associated move-path, so the
```

petrochenkov commented:
r=me after addressing the remaining comments (https://github.com/rust-lang/rust/pull/116300#discussion_r1355119446, https://github.com/rust-lang/rust/pull/116300#discussion_r1357296014) or confirming that they are not worth addressing.
@rustbot author

bors commented:
:umbrella: The latest upstream changes (presumably #116958) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

cjgillot committed:
Use terse form for Fn bound.

cjgillot committed:
Remove redundant checks.

cjgillot committed:
Typo.

bors commented:
:pushpin: Commit c8f33ec35f7dccf9890e9f8438c221ac74c1e07d has been approved by `petrochenkov`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=petrochenkov c8f33ec35f7dccf9890e9f8438c221ac74c1e07d -->
<!-- homu: {"type":"Approved","sha":"c8f33ec35f7dccf9890e9f8438c221ac74c1e07d","approver":"petrochenkov","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:hourglass: Testing commit c8f33ec35f7dccf9890e9f8438c221ac74c1e07d with merge cd674d61790607dfb6faa9d754bd3adfa13aea7c...
<!-- homu: {"type":"BuildStarted","head_sha":"c8f33ec35f7dccf9890e9f8438c221ac74c1e07d","merge_sha":"cd674d61790607dfb6faa9d754bd3adfa13aea7c"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/6620626135/job/17985608047)
Approved by: petrochenkov
Pushing cd674d61790607dfb6faa9d754bd3adfa13aea7c to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"petrochenkov","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/6620626135/job/17985608047"},"merge_sha":"cd674d61790607dfb6faa9d754bd3adfa13aea7c"} -->

rust-timer commented:
Finished benchmarking commit (cd674d61790607dfb6faa9d754bd3adfa13aea7c): [comparison URL](https://perf.rust-lang.org/compare.html?start=f654229c27267334023a22233795b88b75fc340e&end=cd674d61790607dfb6faa9d754bd3adfa13aea7c&stat=instructions:u).

### Overall result: ✅ improvements - no action needed


@rustbot label: -perf-regression



### [Instruction count](https://perf.rust-lang.org/compare.html?start=f654229c27267334023a22233795b88b75fc340e&end=cd674d61790607dfb6faa9d754bd3adfa13aea7c&stat=instructions:u)
This is a highly reliable metric that was used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 0.2%  | [0.1%, 0.2%]   | 5     |
| Regressions ❌ <br /> (secondary)  | 0.2%  | [0.2%, 0.2%]   | 2     |
| Improvements ✅ <br /> (primary)   | -0.7% | [-3.6%, -0.2%] | 77    |
| Improvements ✅ <br /> (secondary) | -1.1% | [-1.8%, -0.4%] | 30    |
| All ❌✅ (primary)                 | -0.7% | [-3.6%, 0.2%]  | 82    |

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=f654229c27267334023a22233795b88b75fc340e&end=cd674d61790607dfb6faa9d754bd3adfa13aea7c&stat=max-rss)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 1.9%  | [0.0%, 2.8%]   | 3     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -3.5% | [-4.8%, -2.9%] | 4     |
| Improvements ✅ <br /> (secondary) | -1.5% | [-4.0%, -0.7%] | 5     |
| All ❌✅ (primary)                 | -1.2% | [-4.8%, 2.8%]  | 7     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=f654229c27267334023a22233795b88b75fc340e&end=cd674d61790607dfb6faa9d754bd3adfa13aea7c&stat=cycles:u)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 1.3%  | [1.3%, 1.3%]   | 1     |
| Regressions ❌ <br /> (secondary)  | 2.2%  | [1.7%, 3.1%]   | 6     |
| Improvements ✅ <br /> (primary)   | -0.9% | [-1.1%, -0.6%] | 4     |
| Improvements ✅ <br /> (secondary) | -     | -              | 0     |
| All ❌✅ (primary)                 | -0.4% | [-1.1%, 1.3%]  | 5     |
</details>

### [Binary size](https://perf.rust-lang.org/compare.html?start=f654229c27267334023a22233795b88b75fc340e&end=cd674d61790607dfb6faa9d754bd3adfa13aea7c&stat=size:linked_artifact)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 0.1%  | [0.1%, 0.2%]   | 7     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -0.1% | [-0.3%, -0.0%] | 8     |
| Improvements ✅ <br /> (secondary) | -     | -              | 0     |
| All ❌✅ (primary)                 | -0.0% | [-0.3%, 0.2%]  | 15    |
</details>

**Bootstrap**: missing data
**Artifact size**: 304.34 MiB -> 304.35 MiB (0.00%)
